internal/tools/terraformlinter/terraform_linter.go

```diff
@@ -0,0 +1,335 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package terraformlinter contains a linter implementation that verifies terraform
+// files against our internal style guide and reports on all violations.
+package terraformlinter
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+)
+
+// Top level terraform types to validate.
+const (
+	tokenTypeResource = "resource"
+	tokenTypeModule   = "module"
+)
+
+// List of valid extensions that can be linted.
+var terraformSelectors = []string{".tf", ".tf.json"}
+
+// Enum of positional locations in order.
+type tokenPosition int32
+
+const (
+	None tokenPosition = iota
+	Leading
+	Provider
+	Ignored
+	Trailing
+)
+
+// tokenAttr defines an individual attribute within a block of terraform.
+type tokenAttr struct {
+	tokenPos        tokenPosition
+	token           hclsyntax.Token
+	trailingNewline bool
+}
+
+// keywords to match on.
+const (
+	attrForEach                = "for_each"
+	attrCount                  = "count"
+	attrProvider               = "provider"
+	attrSource                 = "source"
+	attrProviderProject        = "project"
```

Comment:
Can we enforce the ordering:

```
org
folder
project
```

Reply:
Can be a follow up as I assume thats not a 2 line fix.

Reply:
Your doc doesn't say it needs to be ordered last I checked. I was going to order them but then thought "woohoo I can skip that part!"

I'll get it in a follow up unless I get a bunch of other changes I need to do anyhow.

Reply:
Updated doc. Thanks.

Reply:
Filed https://github.com/abcxyz/pkg/issues/78 and will fix this in there.